### PR TITLE
Implement v2 API

### PR DESCRIPTION
**Pull request by akhilles**
Implements https://github.com/mrhooray/crc-rs/issues/48

### CODE REVIEW

**Comment by akhilles** [comment]
I see a ~15% perf regression in crc64 case, will have to dig into it
```
crc16_update_megabytes  time:   [1.9334 ms 1.9373 ms 1.9415 ms]
                        thrpt:  [491.21 MiB/s 492.27 MiB/s 493.27 MiB/s]
                 change:
                        time:   [+0.8559% +1.5330% +2.1055%] (p = 0.00 < 0.05)
                        thrpt:  [-2.0621% -1.5099% -0.8487%]
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  1 (1.00%) high severe

Benchmarking crc32_update_megabytes: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 8.6s, enable flat sampling, or reduce sample count to 50.
crc32_update_megabytes  time:   [1.6994 ms 1.7092 ms 1.7204 ms]
                        thrpt:  [554.32 MiB/s 557.98 MiB/s 561.18 MiB/s]
                 change:
                        time:   [-0.9138% -0.0424% +0.6526%] (p = 0.92 > 0.05)
                        thrpt:  [-0.6484% +0.0424% +0.9222%]
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe

crc64_update_megabytes  time:   [2.0104 ms 2.0338 ms 2.0605 ms]
                        thrpt:  [462.84 MiB/s 468.92 MiB/s 474.36 MiB/s]
                 change:
                        time:   [+15.056% +16.614% +18.444%] (p = 0.00 < 0.05)
                        thrpt:  [-15.572% -14.247% -13.086%]
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  9 (9.00%) high severe
```

File: src/crc16.rs [hunk 1]
```diff
@@ -1,158 +1,65 @@
...
-    /// digest.write(b"123456789");
-    /// assert_eq!(digest.sum16(), 0x906e);
-    /// ```
-    pub fn new_custom(poly: u16, initial: u16, final_xor: u16, calc: CalcType) -> Digest {
-        let mut rfl: bool = true;
-        if let CalcType::Normal = calc {
-            rfl = false;
-        }
+    fn table_entry(&self, index: u16) -> u16 {
+        unsafe { *self.table.get_unchecked((index & 0xFF) as usize) }
```
[the remark this answers is not in the document]
Comment (akhilles) [reply src/crc16.rs:25]: We can remove this unsafe block. Unfortunately, the compiler isn't able to infer that `(index & 0xFF)` will always be <= 255. So, it's adding an unnecessary bounds check to the array access. However, this cost might be negligible, I can run some benchmarks.

**Comment by akhilles** [comment]
> @akhilles thanks for the PR!
> 
> Looks like the focus is on the API & improved compatibility rather than internal code dedup between 16/32/64 widths - is this understanding correct?

Yeah, the main blocker for eliminating the code dedup is the ability to use generics in const functions. Once that's supported, we can make that change without breaking the API.

> What's your thought on the release plan as this requires Rust beta? My understanding is 1) folks that require stable wouldn't be able to use this immediately; 2) later adding stable channel support wouldn't be a breaking change so we can stay on v2

I can revert to using macros in `make_table`, which works fine in stable rustc. Probably want to avoid increasing MSRV unless absolutely necessary.

**Comment by akhilles** [comment]
One downside of using macros in `make_table` is it drastically increases compile times (from ~1s to ~20s). Another option we have is to manually unroll all the loops. It'll make the `table.rs` file huge but will compile fast and work on stable rustc.

File: src/crc16.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (akhilles) [reply src/crc16.rs]: didn't observe any performance impact, removed the unsafe code

**Comment by mrhooray** [comment]
Thanks @akhilles 

Do you think the `core::hash::Hasher` trait is worth implementing, remembered that there were requests for it.

**Pull request merged**